lib/generate/consents.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+module Generate
+  class Consents
+    attr_reader :organisation, :programme
+
+    def initialize(
+      organisation:,
+      programme: nil,
+      session: nil,
+      refused: 0,
+      given: 0,
+      given_needs_triage: 0
+    )
+      @organisation = organisation
+      @programme = programme || organisation.programmes.sample
+      @session = session
+      @refused = refused
+      @given = given
+      @given_needs_triage = given_needs_triage
+    end
+
+    def call
+      create_consent_refused(@refused)
+      create_consent_given(@given)
+      create_consent_given_needs_triage(@given_needs_triage)
+    end
+
+    def self.call(...) = new(...).call
+
+    private
+
+    def patients
+      (@session.presence || organisation)
+        .patients
+        .includes(:parents, :consents, consents: :parent)
+        .in_programmes([programme])
+        .select { it.consents.empty? && it.parents.any? }
+    end
+
+    def random_patients(count)
+      patients
+        .shuffle
+        .take(count)
+        .tap do
+          if it.size < count
+            raise "Not enough patients without consent and with parents to generate consents"
+          end
+        end
+    end
+
+    def session_for(patient)
+      patient
+        .sessions
+        .eager_load(:location)
+        .merge(Location.school)
+        .has_programme(programme)
+        .sample
+    end
+
+    def create_consent_refused(count)
```

Comment:
This method along with `create_consent_given` are very similar, I wonder if they could be refactored in to a single method?